### ISSUE TITLE
hipchat: module fails when API return 204 HTTP status code

### ISSUE THREAD

**Issue by llernestal** [ansible/ansible-modules-extras#2143]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

Hipchat
##### ANSIBLE VERSION

```
ansible 2.1.0.0 (stable-2.1 c7cae3b08c) last updated 2016/05/02 11:16:23 (GMT +000)
  lib/ansible/modules/core: (detached HEAD babb97e473) last updated 2016/05/02 11:16:30 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 1d180c7707) last updated 2016/05/02 11:16:34 (GMT +000)
```
##### SUMMARY

Module fails when Hipchat REST API v2 returns a non `200` HTTP status code
##### STEPS TO REPRODUCE

```
- name: Notify hipchat
  hipchat:
    api: "https://api.hipchat.com/v2"
    color: gray
    room: "{{ item.room }}"
    token: "{{ item.token }}"
    msg: "hello world"
    msg_format: html
  with_items: "{{ hipchat }}"
```

##### EXPECTED RESULTS

A successful return from module when Hipchat REST API v2 returns a 2xx status code
##### ACTUAL RESULTS

```
failed: [xxx.xxx.xxx.xxx] (item={u'token': u'OAUTH_TOKEN', u'room': u'Bot chat'}) => {"failed": true, "invocation": {"module_args": {"api": "https://api.hipchat.com/v2", "color": "gray", "msg": "hello world", "msg_format": "html", "msg_from": "Ansible", "notify": true, "room": "Bot chat", "token": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER", "validate_certs": true}, "module_name": "hipchat"}, "item": {"room": "Bot chat", "token": "OAUTH_TOKEN"}, "msg": "failed to send message, return status=204"}
```

**Comment by alikins** [comment]
Yup, looks like the module only considers a 200 a success.
https://github.com/ansible/ansible-modules-extras/blob/devel/notification/hipchat.py#L157

https://www.hipchat.com/docs/apiv2/method/send_room_notification does show 204 as the expected return code  (Even though 204 is missing from https://developer.atlassian.com/hipchat/guide/hipchat-rest-api/api-response-codes at the moment)

**Comment by llernestal** [comment]
The module works for me when running:

```
ansible 2.0.0.2 (stable-2.0.0.1 3b5d7400de) last updated 2016/05/02 14:17:53 (GMT +000)
  lib/ansible/modules/core: (detached HEAD ce6619bf5d) last updated 2016/05/02 14:18:02 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 29af26884e) last updated 2016/05/02 14:18:08 (GMT +000)
```

but, of course, with room name URL-encoded:

```
- name: Notify hipchat
  hipchat:
    api: "https://api.hipchat.com/v2"
    color: gray
    room: "{{ item.room|urlencode }}"
    token: "{{ item.token }}"
    msg: "hello world"
    msg_format: html
  with_items: "{{ hipchat }}"
```

Isn't that strange since https://github.com/ansible/ansible-modules-extras/blob/stable-2.0.0.1/notification/hipchat.py#L158 also says only `200` code is accepted? Or am I missing something here?

**Comment by alikins** [comment]
> Isn't that strange since https://github.com/ansible/ansible-modules-extras/blob/stable-2.0.0.1/notification/hipchat.py#L158 also says only 200 code is accepted? Or am I missing something here?

Likely a result of the changes related to fixes for #15555.
84ee00bb4246bae43e4fdd149b72e9f1b5693c1e or 7f206fdfe8b2374ea289f6140d78ffc3d4d56745 or changes/fixes in fetch_url
( https://github.com/ansible/ansible/commit/ed35e8bce21180529e72f5bbe1d853b67ddac870 or https://github.com/ansible/ansible/commit/ 4647e8b74e614e5ed82becd0eb502d79d10d2b13 etc).

2.0.2 was likely ignoring the 204 and returning a 200 from fetch_url.
